### DISCUSSION TITLE
Is it possible to add parser information to a file's front matter data?

### DISCUSSION THREAD

**Discussion by pspeter3** [11ty/eleventy#1468] [Q&A]
Eg is it possible to add a `toc` or `links` property to the front matter data based on the parser? This could be achieved by parsing the content in `eleventyComputed` but that seems comparatively wasteful.

**Comment by Ryuno-Ki** [comment]
Hm, couldn't you use a [transform](https://www.11ty.dev/docs/config/#transforms) for that?
Why do you need the information in the front-matter?

**Reply by pspeter3** [reply]
I want to use the information as part of rendering templates. My specific use cases are adding to the TOC since there are headings from data and not in the markdown or rendering back links similar to Eleventy Garden. In the second case you can imagine filtering all the posts which contain a link to the current post.

**Reply by Ryuno-Ki** [reply]
Hm, I could imagine, placing a HTML token as anchor in the Markdown file.
At the point when a transform is processed, the Markdown (and data) were already transformed.
That is, you know about the existing `<h1>`-`<h6>`.
That would allow you to build a Table Of Contents and inject it into the place the HTML comments was.
(I believe, WordPress did something similar).

Regarding the links it might be a bit trickier. You perhaps replace the node name from `<a>` to `<span>`? But then it can't have block elements as children :thinking:

**Reply by pspeter3** [reply]
I actually don't want to change any of the output to be clear. I just want to be able to use the parser hooks to create an array of links. I could parse the content again but as I said, it seems wasteful to do a second pass if I don't have to.

**Reply by pspeter3** [reply]
Another alternative to my solution could be just to make a data file which just globs and parses all the markdown there.

**Reply by Ryuno-Ki** [reply]
You wouldn't be far from a [collection](https://www.11ty.dev/docs/collections/#getfilteredbytag(-tagname-)) than?

**Reply by pspeter3** [reply]
@Ryuno-Ki I'm unsure what you mean.

**Reply by pspeter3** [reply]
Oh are you proposing that for each file, I make a collection which is the files that link to it?

**Reply by Ryuno-Ki** [reply]
I'm not quite sure to got you either :sweat_smile: 

> Another alternative to my solution could be just to make a data file which just globs and parses all the markdown there.

For globbing, there are already options. For example, collections. You could look into whether those get close to what you want before reinventing the wheel.

> I just want to be able to use the parser hooks to create an array of links.

Links to what?

My understanding right now is, that you have a blog post. That is part of a multi-part article series about a certain topic. Perhaps even a longread. So there shall be a ToC at the beginning to prepare the reader about the content below. At the bottom, you want to link to other articles in that series.

Those article-series could share a tag for example, which would allow the creation of a collection.
The collection could be [paginated](https://www.11ty.dev/docs/pagination/#paging-a-collection) to generate the prev- and next links.

But perhaps I got the use case wrong. What is this Eleventy Garden?

**Reply by pspeter3** [reply]
[Eleventy Garden](https://github.com/binyamin/eleventy-garden) is an example of someone trying to implement Roam In Eleventy. Imagine that you have the following files:

**alpha.md**
```md
---
name: Alpha
---
The beginning of the site.
```

**beta.md**
```
---
name: Beta
---
The second page in the site. Comes after [Alpha](/alpha/)
```

**omega.md**
```md
---
name: Omega
---
The last page in the site. Comes after [Alpha](/alpha/) and [Beta](/beta/)
```

I would want the following front matters:
```js
[
  { "permalink": "/alpha/", "links": [], /* ... */ }
  { "permalink": "/beta/", "links": ["/alpha/"], /* ... */ }
  { "permalink": "/alpha/", "links": ["/alpha/", "/beta/"], /* ... */ }
]
```

That way I can show that Beta and Omega link to alpha on the alpha page. I'm trying to build backlinks via the `link` property.

**Reply by Ryuno-Ki** [reply]
How is the parser supposed to know the order in your opinion?

**Reply by pspeter3** [reply]
It's not supposed to know the order. I just want the parser to expose that Omega links to Alpha & Beta.

**Comment by zachleat** [comment]
Interesting—are there any hooks you can use via a markdown-it plugin? See https://www.11ty.dev/docs/languages/markdown/#optional-set-your-own-library-instance

**Reply by pspeter3** [reply]
You would need to add a custom rule that basically produces a side effect. Essentially the equivalent of creating a linter or transform in Eleventy. The API is certainly easier in `marked` which has events for each node but you could make it work with `markdown-it` as well. Creating an API to connect the rule to the Eleventy data is the hard question.

**Reply by pspeter3** [reply]
For example you could use the callback on the markdown-it-anchor to build the TOC if you wanted to https://github.com/valeriangalliat/markdown-it-anchor

**Comment by kkgthb** [comment]
I think I just found a workaround, if I'm understanding this correctly, @pspeter3.  I just added both a `pages.11tydata.js` and a `pages.11tydata.json` file to a `pages` folder.  The JavaScript took care of things w/ complex variables, like stripping `pages` out of `permalink` even if `permalink` was set in the `.md` file's front matter.  The only thing it wouldn't let me set was `tags`.  So I did `{"tags": "pages"}` in the JSON and they work together like a charm as of late January, 2021.
